outline/shadowsocks/config.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2022 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shadowsocks
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Jigsaw-Code/outline-go-tun2socks/outline/internal/encoding/utf8"
+)
+
+// [Exported] Config represents a shadowsocks server configuration.
+type Config struct {
+	Host       string
+	Port       int
+	Password   string
+	CipherName string
+	Prefix     []byte
+}
+
+// An internal data structure to be used by JSON deserialization.
+// Must match the ShadowsocksSessionConfig interface defined in Outline Client.
+type configJSON struct {
+	Host     string `json:"host"`
+	Port     uint16 `json:"port"`
+	Password string `json:"password"`
+	Method   string `json:"method"`
+	Prefix   string `json:"prefix"`
+}
+
+// [Exported] ParseConfigFromJSON parses a JSON string `in` as a Config object.
+// The JSON string `in` must match the ShadowsocksSessionConfig interface
+// defined in Outline Client.
+func ParseConfigFromJSON(in string) (*Config, error) {
```

Comment:
I don't see the point of having this exported method.
Why not have a `NewClientFromJSON(configJson string)`?

That Config type is not a good abstraction, and we should avoid it.

Reply:
I'm trying to separate the concept of `config` and `client`. JSON is only related `config`, client doesn't need to know how the `config` is created.

But if the config interface is not well defined now, we may introduce `NewClientFromJSON` for now. Though I think we need to define it in SDK.

Reply:
That makes sense, so I guess the code here will be moved to Outline-Apps repo? If this is the case, we can directly create the client (different than the SDK's client) from JSON.